File: src/unicode.ts
// Vendored Unicode 15.1.0 tables (scripts/gen_unicode_tables.mjs) plus the
// grapheme cluster segmenter (UAX #29), full case folding (C+F mappings),
// and the property lookups the screening pipeline needs. Unicode data is
// vendored and pinned by digest per spec section 2; NFKC uses the host
// runtime's String.prototype.normalize (ICU) and is the only non-vendored
// table — recorded in STATUS.md.

import tables from './unicode_15_1.json' with { type: 'json' };

export type Ranges = [number, number][];

export const UNICODE_VERSION: string = tables.unicode;
export const UNICODE_TABLE_SHA256 = '9442ce4c0e8e131200034e90ca568dc0b6255418db0fecdf29a83a78d73ee999';

const gcb = tables.gcb as unknown as Record<string, Ranges>;
const extPict = tables.extended_pictographic as unknown as Ranges;
const dicpRanges = tables.default_ignorable as unknown as Ranges;
const wsRanges = tables.white_space as unknown as Ranges;
const catLetter = tables.categories.letter as unknown as Ranges;
const catNumber = tables.categories.number as unknown as Ranges;
const catPc = tables.categories.connectorPunctuation as unknown as Ranges;
const casefoldMap = tables.casefold as unknown as Record<string, number[]>;

function inRanges(ranges: Ranges, cp: number): boolean {
  let lo = 0;
  let hi = ranges.length - 1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    const r = ranges[mid]!;
    if (cp < r[0]) hi = mid - 1;
    else if (cp > r[1]) lo = mid + 1;
    else return true;
  }
  return false;
}

export type Gcb =
  | 'Other' | 'Prepend' | 'CR' | 'LF' | 'Control' | 'Extend'
  | 'Regional_Indicator' | 'SpacingMark' | 'L' | 'V' | 'T' | 'LV' | 'LVT' | 'ZWJ';

export function gcbOf(cp: number): Gcb {
  for (const prop of ['Prepend', 'CR', 'LF', 'Control', 'Extend', 'Regional_Indicator', 'SpacingMark', 'L', 'V', 'T', 'LV', 'LVT', 'ZWJ'] as const) {
    const r = gcb[prop];
    if (r && inRanges(r, cp)) return prop;
  }
  return 'Other';
}

export function isExtendedPictographic(cp: number): boolean {
  return inRanges(extPict, cp);
}

export function isDefaultIgnorable(cp: number): boolean {
  return inRanges(dicpRanges, cp);
}

export function isWhiteSpace(cp: number): boolean {
  return inRanges(wsRanges, cp);
}

export function isTokenChar(cp: number): boolean {
  return inRanges(catLetter, cp) || inRanges(catNumber, cp) || inRanges(catPc, cp);
}

// Full default case folding: CaseFolding.txt status C+F, vendored.
export function caseFoldCp(cp: number): number[] {
  const m = casefoldMap[String(cp)];
  return m ?? [cp];
}

export function caseFoldString(s: string): string {
  let out = '';
  for (const ch of s) {
    for (const cp of caseFoldCp(ch.codePointAt(0)!)) out += String.fromCodePoint(cp);
  }
  return out;
}

// --- UAX #29 extended grapheme cluster segmentation -----------------------

function isGcbControl(g: Gcb): boolean {
  return g === 'Control' || g === 'CR' || g === 'LF';
}

// Returns cluster boundaries: an array of [start,end) index pairs over cps.
export function graphemeClusters(cps: number[]): [number, number][] {
  const n = cps.length;
  const out: [number, number][] = [];
  if (n === 0) return out;
  const g = cps.map(gcbOf);
  let start = 0;
  // State for GB11: index of the most recent ZWJ that is preceded by
  // ExtPict Extend* — i.e. an ExtPict char followed only by Extend chars.
  let extPictBeforeZwRun = false; // set when current position is after ZWJ with GB11 left context
  let riRun = g[0] === 'Regional_Indicator' ? 1 : 0; // consecutive RI count ending at i-1

  for (let i = 1; i < n; i++) {
    const prev = g[i - 1]!;
    const cur = g[i]!;
    let brk = true; // GB999 default

    if (prev === 'CR' && cur === 'LF') brk = false; // GB3
    else if (isGcbControl(prev) || isGcbControl(cur)) brk = true; // GB4/GB5
    else if (prev === 'L' && (cur === 'L' || cur === 'V' || cur === 'LV' || cur === 'LVT')) brk = false; // GB6
    else if ((prev === 'LV' || prev === 'V') && (cur === 'V' || cur === 'T')) brk = false; // GB7
    else if ((prev === 'LVT' || prev === 'T') && cur === 'T') brk = false; // GB8
    else if (cur === 'Extend' || cur === 'ZWJ') brk = false; // GB9
    else if (cur === 'SpacingMark') brk = false; // GB9a
    else if (prev === 'Prepend') brk = false; // GB9b
    else if (prev === 'ZWJ' && extPictBeforeZwRun && isExtendedPictographic(cps[i]!)) brk = false; // GB11
    else if (prev === 'Regional_Indicator' && cur === 'Regional_Indicator' && riRun % 2 === 1) brk = false; // GB12/13

    if (brk) {
      out.push([start, i]);
      start = i;
    }
    // update GB11/GB12 state for position i (as "previous" for i+1)
    if (cur === 'ZWJ') {
      // look back: cps[i-1-...] Extend* then ExtPict
      let j = i - 1;
      while (j >= start && g[j] === 'Extend') j--;
      extPictBeforeZwRun = j >= 0 && isExtendedPictographic(cps[j]!);
    } else if (cur !== 'Extend') {
      extPictBeforeZwRun = false;
    }
    riRun = cur === 'Regional_Indicator' ? (prev === 'Regional_Indicator' ? riRun + 1 : 1) : 0;
  }
  out.push([start, n]);
  return out;
}

File: python/lexsieve/unicode.py
"""Vendored Unicode 15.1.0 tables (scripts/gen_unicode_tables.mjs) plus the
grapheme cluster segmenter (UAX #29), full case folding (C+F mappings), and
the property lookups the screening pipeline needs. Unicode data is vendored
and pinned by digest per spec section 2; NFKC uses the host runtime's
unicodedata (UCD 15.0) and is the only non-vendored table — recorded in
STATUS.md.
"""

import json
import os

_path = os.path.join(os.path.dirname(__file__), "unicode_15_1.json")
with open(_path, "r", encoding="utf-8") as _f:
    _tables = json.load(_f)

UNICODE_VERSION = _tables["unicode"]
UNICODE_TABLE_SHA256 = "9442ce4c0e8e131200034e90ca568dc0b6255418db0fecdf29a83a78d73ee999"

_gcb = _tables["gcb"]
_ext_pict = _tables["extended_pictographic"]
_dicp_ranges = _tables["default_ignorable"]
_ws_ranges = _tables["white_space"]
_cat_letter = _tables["categories"]["letter"]
_cat_number = _tables["categories"]["number"]
_cat_pc = _tables["categories"]["connectorPunctuation"]
_casefold_map = _tables["casefold"]


def _in_ranges(ranges, cp):
    lo, hi = 0, len(ranges) - 1
    while lo <= hi:
        mid = (lo + hi) >> 1
        r = ranges[mid]
        if cp < r[0]:
            hi = mid - 1
        elif cp > r[1]:
            lo = mid + 1
        else:
            return True
    return False


_GCB_ORDER = (
    "Prepend", "CR", "LF", "Control", "Extend", "Regional_Indicator",
    "SpacingMark", "L", "V", "T", "LV", "LVT", "ZWJ",
)


def gcb_of(cp):
    for prop in _GCB_ORDER:
        r = _gcb.get(prop)
        if r and _in_ranges(r, cp):
            return prop
    return "Other"


def is_extended_pictographic(cp):
    return _in_ranges(_ext_pict, cp)


def is_default_ignorable(cp):
    return _in_ranges(_dicp_ranges, cp)


def is_white_space(cp):
    return _in_ranges(_ws_ranges, cp)


def is_token_char(cp):
    return (
        _in_ranges(_cat_letter, cp)
        or _in_ranges(_cat_number, cp)
        or _in_ranges(_cat_pc, cp)
    )


def case_fold_cp(cp):
    """Full default case folding: CaseFolding.txt status C+F, vendored."""
    return _casefold_map.get(str(cp), [cp])


def case_fold_string(s):
    out = []
    for ch in s:
        for cp in case_fold_cp(ord(ch)):
            out.append(chr(cp))
    return "".join(out)


# --- UAX #29 extended grapheme cluster segmentation ------------------------

_GCB_CONTROLS = frozenset({"Control", "CR", "LF"})


def grapheme_clusters(cps):
    """Cluster boundaries: list of [start,end) index pairs over cps."""
    n = len(cps)
    out = []
    if n == 0:
        return out
    g = [gcb_of(c) for c in cps]
    start = 0
    # GB11 state: set when current position is after a ZWJ preceded by
    # ExtPict Extend*.
    ext_pict_before_zw_run = False
    ri_run = 1 if g[0] == "Regional_Indicator" else 0

    for i in range(1, n):
        prev, cur = g[i - 1], g[i]
        brk = True  # GB999 default
        if prev == "CR" and cur == "LF":
            brk = False  # GB3
        elif prev in _GCB_CONTROLS or cur in _GCB_CONTROLS:
            brk = True  # GB4/GB5
        elif prev == "L" and cur in ("L", "V", "LV", "LVT"):
            brk = False  # GB6
        elif prev in ("LV", "V") and cur in ("V", "T"):
            brk = False  # GB7
        elif prev in ("LVT", "T") and cur == "T":
            brk = False  # GB8
        elif cur == "Extend" or cur == "ZWJ":
            brk = False  # GB9
        elif cur == "SpacingMark":
            brk = False  # GB9a
        elif prev == "Prepend":
            brk = False  # GB9b
        elif prev == "ZWJ" and ext_pict_before_zw_run and is_extended_pictographic(cps[i]):
            brk = False  # GB11
        elif prev == "Regional_Indicator" and cur == "Regional_Indicator" and ri_run % 2 == 1:
            brk = False  # GB12/13

        if brk:
            out.append([start, i])
            start = i
        if cur == "ZWJ":
            j = i - 1
            while j >= start and g[j] == "Extend":
                j -= 1
            ext_pict_before_zw_run = j >= 0 and is_extended_pictographic(cps[j])
        elif cur != "Extend":
            ext_pict_before_zw_run = False
        ri_run = (
            (ri_run + 1) if prev == "Regional_Indicator" else 1
        ) if cur == "Regional_Indicator" else 0

    out.append([start, n])
    return out
